src/types/Message.ts

```diff
@@ -0,0 +1,184 @@
+  content: {
+    /** Optional machine-readable type of the rejection. */
+    name?: string
+    /** Optional human-readable description of the rejection. */
+    message?: string
+  }
+  type: 'reject'
+}
+
+export interface ISubmitTerms extends IMessageBodyBase {
+  content: ITerms
+  type: 'submit-terms'
+}
+
+export interface IRequestAttestationContent {
+  credential: ICredential
+  quote?: IQuoteAgreement
+}
+
+export interface IRequestAttestation extends IMessageBodyBase {
+  content: IRequestAttestationContent
+  type: 'request-attestation'
+}
+
+export interface IRequestPaymentContent {
+  // Same as the `rootHash` value of the `'request-attestation'` message */
+  claimHash: string
+}
+
+export interface IRequestPayment extends IMessageBodyBase {
+  content: IRequestPaymentContent
+  type: 'request-payment'
+}
+
+export interface ISubmitAttestationContent {
+  attestation: IAttestation
+}
+
+export interface ISubmitAttestation extends IMessageBodyBase {
+  content: ISubmitAttestationContent
+  type: 'submit-attestation'
+}
+
+export interface IRejectAttestation extends IMessageBodyBase {
+  content: ICredential['rootHash']
+  type: 'reject-attestation'
+}
+
+export interface IRequestCredentialContent {
+  cTypes: Array<{
+    cTypeHash: CTypeHash
+    trustedAttesters?: DidUri[]
+    requiredProperties?: string[]
+  }>
+  challenge?: string
+}
+
+export interface IRejectTermsContent {
+  claim: PartialClaim
+  legitimations: ICredential[]
+  delegationId?: IDelegationNode['id']
+}
+
+
+
+export interface ISubmitCredential extends IMessageBodyBase {
+  content: ICredentialPresentation[]
+  type: 'submit-credential'
+}
+
+export interface IRequestCredential extends IMessageBodyBase {
+  content: IRequestCredentialContent
+  type: 'request-credential'
+}
+
+export interface IConfirmPaymentContent {
+  // Same as the `rootHash` value of the `'request-attestation'` message
+  claimHash: string
+  // Hash of the payment transaction */
+  txHash: string
+  // hash of the block which includes the payment transaction
+  blockHash: string
+}
+
+export interface IConfirmPayment extends IMessageBodyBase {
+  content: IConfirmPaymentContent
+  type: 'confirm-payment'
+}
+
+export type MessageBody =
+  | IError
+  | IReject
+  | ISubmitTerms
+  | IRequestAttestation
+  | IRequestPayment
+  | IConfirmPayment
+  | ISubmitAttestation
+  | IRejectAttestation
+  | IRequestCredential
+  | ISubmitCredential
+
+/**
+ * - `body` - The body of the message, see [[MessageBody]].
+ * - `createdAt` - The timestamp of the message construction.
+ * - `sender` - The DID of the sender.
+ * - `receiver` - The DID of the receiver.
+ * - `messageId` - The message id.
+ * - `receivedAt` - The timestamp of the message reception.
+ * - `inReplyTo` - The id of the parent-message.
+ * - `references` - The references or the in-reply-to of the parent-message followed by the message-id of the parent-message.
+ */
+export interface IMessage {
```

Comment:
can we make this type generic over MessageBody? Default type should be a more general version of MessageBody, e.g.
```ts
{
  type: string,
  content: unknown
}

Reply:
Sorry, I do not get this point.

Reply:
Something like this:

```
interface IMessageBodyBase<Content = unknown> {
  content: Content
  type: string
}
```

Reply:
Wrong type...

```ts
export interface IMessage<Body: IMessageBodyBase> {
  body: Body
  createdAt: number
  sender: DidUri
  receiver: DidUri
  messageId?: string
  receivedAt?: number
  inReplyTo?: IMessage['messageId']
  references?: Array<IMessage['messageId']>
}
```

Reply:
yes, but IMessageBodyBase itself shouldn't be generic, or at least in a way that makes sure that the type string and content type remain linked. Could be achieved as follows, for example:
```ts
interface IMessageBodyBase<T extends string = string, C = unknown> {
  type: T
  content: C
}

type MessageBodyA = IMessageBodyBase<"A", number>

type MessageBodyB = IMessageBodyBase<"B", boolean>

type MessageBodies = MessageBodyA | MessageBodyB

function doStuff(x: MessageBodies) {
  console.log(x)
}

// complains
doStuff({
  type: "B",
  content: 1,
})
```

Reply:
I thought having `IMessageBodyBase` generic would be nice. But seeing your code @rflechtner i don't see that we gain on readability or conciseness

Reply:
I like it, it does make the message type definitions a little more concise, doesn't it?